### PR TITLE
feat(noise): add support to Waku Payload V2

### PR DESCRIPTION
**Pull request by s1fr0**
This PR introduces data structures and procedures to process Waku Message Payload with version 2. 

Specifically, it adds:
- Encode/decode payload functions
- Some unit tests

Implementation of Waku Payload with version 2 partially follows [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/#specification) (RFC will be updated accordingly once this PR is merged).

This PR extends https://github.com/status-im/nim-waku/pull/931 (which in turn extends https://github.com/status-im/nim-waku/pull/930) and partially addresses https://github.com/status-im/nim-waku/issues/881.

### CODE REVIEW

**Comment by staheri14** [comment]
I see this PR and PR #931 have overlaps e.g., they both have the definition of `decryptNoisePublicKey`, so I think it makes sense to wait for PR#931 to get merged then I'll review this one.

**Comment by s1fr0** [comment]
@staheri14 Ready for review

File: waku/v2/protocol/waku_noise/noise.nim
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:353]:
```suggestion
  if handshakeMessageLen > uint8.high:
```

File: waku/v2/protocol/waku_noise/noise.nim
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:349]:
```suggestion
  i += 1
```

File: waku/v2/protocol/waku_noise/noise.nim
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:356]:
```suggestion
  i += 1
```

File: waku/v2/protocol/waku_noise/noise.nim
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:360]:
```suggestion
    # In handshakeMessage we accumulate the deserialized Noise Public keys read
```

**Review by staheri14 — APPROVED** [review]
Looks good to me!  please see my comments inline.

**Comment by staheri14** [comment]
@s1fr0  it seems the `transport-message-len-len` is removed from the serialization, right?

**Comment by s1fr0** [comment]
Thanks @staheri14. Yes, as per PR description, RFC will be updated accordingly once this PR is merged. As I was expecting, minor RFC details changed while implementing.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:326]: This will be done in next PR (https://github.com/status-im/nim-waku/pull/933), when the list of supported protocol IDs and Noise handshakes is introduced.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Add to to an extra cast to int. Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:317]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:283]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7.

However, would it better to keep track of specific returned error objects, e.g. `NoisePublicKeyError`, `NoiseMalformedHandshake` etc. ? It seems that with `err`, functions which call a proc with a `Result` output, will loose the information regarding the type of errors returned by the called proc.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (staheri14) [reply waku/v2/protocol/waku_noise/noise.nim]: Good question, as per Nim style guide, [Error handling section](https://status-im.github.io/nim-style-guide/errors.html#error-handling) 
> Because of the controversies and changing landscape, the preference for Status projects is to avoid the use of exceptions unless specially motivated, if only to maintain consistency and simplicity.

**Comment by staheri14** [comment]
I see this PR is not merged yet, shall I review #933?

**Comment by oskarth** [comment]
@jm-clius are you able to review this one as another sanity check?

**Comment by s1fr0** [comment]
@staheri14 I was waiting another approval before merging this PR into master. After that, it is easier to see only new changes introduced by https://github.com/status-im/nim-waku/pull/933. In case you want to review https://github.com/status-im/nim-waku/pull/933 already, I proceeded with a rebase on this PR's current branch and added the necessary changes based on past related PRs reviewers' comments (i.e. add documentation, camel case, returns, etc.).

**Comment by staheri14** [comment]
Thanks @s1fr0 for rebasing the PR! However, as there might be changes after @jm-clius's review, I will review #933 after that.

File: waku/v2/node/waku_payload.nim
Comment (jm-clius) [inline waku/v2/node/waku_payload.nim:85]: As an interim step I think having specific `decode` and `encode` functions for version `2` is ok, but we must merge this logic with [the existing `decodePayload` and `encode` procs](https://github.com/status-im/nim-waku/blob/430f2ecd6720d5835c430e310ffb06ce27f840c7/waku/v2/node/waku_payload.nim#L39-L80) if possible. I think that can be done by extending `KeyInfo` and having the return type of `decodePayload` be a case object rather than the version 1 specific `DecodedPayload`.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (jm-clius) [inline waku/v2/protocol/waku_noise/noise.nim:14]: Should go under `std` I think.

**Review by jm-clius — APPROVED** [review]
LGTM, thanks! Left a comment on having a clean `encode` and `decode` interface for all payload versions in `waku_payload`, but can be addressed in subsequent PR.

File: waku/v2/protocol/waku_noise/noise.nim
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nim-waku/commit/76494c2a77672c329567f32b5e18bf15319e9b25

File: waku/v2/node/waku_payload.nim
Comment (jm-clius) [reply waku/v2/node/waku_payload.nim:85]: I agree that this can be a separate PR. Thanks!

Comment (s1fr0) [reply waku/v2/node/waku_payload.nim:85]: To keep track of this, I created the following issue https://github.com/status-im/nim-waku/issues/942

**Pull request merged**
